src/modules/time_persistor/TimePersistor.cpp

```diff
@@ -0,0 +1,177 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2024 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "TimePersistor.hpp"
+
+#include <errno.h>
+#include <px4_platform_common/log.h>
+
+TimePersistor::TimePersistor() :
+	ModuleParams(nullptr)
+{
+}
+
+TimePersistor::~TimePersistor()
+{
+	fclose(_file);
+}
+
+int TimePersistor::init()
+{
+	if ((_file = fopen(TIME_FILE_PATH, "r+")) == nullptr) {
+		_file = fopen(TIME_FILE_PATH, "w+");
+	}
+
+	if (!_file) {
+		PX4_ERR("error opening the time file");
+		return PX4_ERROR;
+	}
+
+	time_t t;
+
+	if (read_time(&t) == PX4_OK) {
+		struct timespec ts;
+		ts.tv_sec = t;
+		ts.tv_nsec = 0;
+
+		px4_clock_settime(CLOCK_REALTIME, &ts);
```

Comment:
Perhaps we could add some basic checks?
- Time should not be set backwards.
- Time should not be set before when it was compiled (if available).
- Time should not be set forward by more than 1y?

Reply:
Thinking some more about this, I think we should centralize this because we have observed similar issues with using GPS time as systemtime, thus we should make a `px4_clock_unstrusted_settime()` function or similar that implements various sanity checks to prevent setting abnormal time.